File: packages/shared/src/currency.ts
import { z } from 'zod';

/**
 * Supported currencies and their minor-unit exponent (ISO 4217).
 * Amounts are always stored as integers in the minor unit:
 * 1.500 KWD → 1500 fils, 12.34 USD → 1234 cents, 500 JPY → 500.
 *
 * Mirrored in Rust (`pos_core::currency`) and pinned by `contracts/currencies.json`.
 */
export const CURRENCIES = {
  AED: { exponent: 2, symbol: 'د.إ' },
  BHD: { exponent: 3, symbol: '.د.ب' },
  EGP: { exponent: 2, symbol: 'E£' },
  EUR: { exponent: 2, symbol: '€' },
  GBP: { exponent: 2, symbol: '£' },
  IQD: { exponent: 3, symbol: 'ع.د' },
  JOD: { exponent: 3, symbol: 'د.ا' },
  JPY: { exponent: 0, symbol: '¥' },
  KWD: { exponent: 3, symbol: 'د.ك' },
  MYR: { exponent: 2, symbol: 'RM' },
  OMR: { exponent: 3, symbol: 'ر.ع.' },
  QAR: { exponent: 2, symbol: 'ر.ق' },
  SAR: { exponent: 2, symbol: 'ر.س' },
  TND: { exponent: 3, symbol: 'د.ت' },
  USD: { exponent: 2, symbol: '$' },
} as const satisfies Record<string, { exponent: 0 | 2 | 3; symbol: string }>;

export type CurrencyCode = keyof typeof CURRENCIES;

export const CURRENCY_CODES = Object.keys(CURRENCIES) as [CurrencyCode, ...CurrencyCode[]];

export const CurrencyCodeSchema = z.enum(CURRENCY_CODES);

export function currencyExponent(code: CurrencyCode): number {
  return CURRENCIES[code].exponent;
}

/**
 * Exact exchange rate as a rational number: 1 major unit of `base` =
 * `numerator / denominator` major units of `quote`. Rationals avoid float drift.
 */
export const ExchangeRateSchema = z.object({
  base: CurrencyCodeSchema,
  quote: CurrencyCodeSchema,
  numerator: z.int().positive(),
  denominator: z.int().positive(),
});
export type ExchangeRate = z.infer<typeof ExchangeRateSchema>;

File: crates/pos-core/src/currency.rs
//! Supported currencies (ISO 4217) and their minor-unit exponents.

use serde::{Deserialize, Serialize};

/// Variants are the ISO 4217 codes verbatim so serde round-trips them unchanged.
#[allow(clippy::upper_case_acronyms)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
pub enum CurrencyCode {
    AED,
    BHD,
    EGP,
    EUR,
    GBP,
    IQD,
    JOD,
    JPY,
    KWD,
    MYR,
    OMR,
    QAR,
    SAR,
    TND,
    USD,
}

impl CurrencyCode {
    pub const ALL: [CurrencyCode; 15] = [
        CurrencyCode::AED,
        CurrencyCode::BHD,
        CurrencyCode::EGP,
        CurrencyCode::EUR,
        CurrencyCode::GBP,
        CurrencyCode::IQD,
        CurrencyCode::JOD,
        CurrencyCode::JPY,
        CurrencyCode::KWD,
        CurrencyCode::MYR,
        CurrencyCode::OMR,
        CurrencyCode::QAR,
        CurrencyCode::SAR,
        CurrencyCode::TND,
        CurrencyCode::USD,
    ];

    /// Number of minor-unit digits: 1.500 KWD = 1500 fils → 3.
    pub const fn exponent(self) -> u32 {
        use CurrencyCode::*;
        match self {
            JPY => 0,
            AED | EGP | EUR | GBP | MYR | QAR | SAR | USD => 2,
            BHD | IQD | JOD | KWD | OMR | TND => 3,
        }
    }

    pub fn as_str(self) -> &'static str {
        use CurrencyCode::*;
        match self {
            AED => "AED",
            BHD => "BHD",
            EGP => "EGP",
            EUR => "EUR",
            GBP => "GBP",
            IQD => "IQD",
            JOD => "JOD",
            JPY => "JPY",
            KWD => "KWD",
            MYR => "MYR",
            OMR => "OMR",
            QAR => "QAR",
            SAR => "SAR",
            TND => "TND",
            USD => "USD",
        }
    }
}

/// Exact rate: 1 major unit of `base` = `numerator / denominator` major units of `quote`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct ExchangeRate {
    pub base: CurrencyCode,
    pub quote: CurrencyCode,
    pub numerator: u64,
    pub denominator: u64,
}
